trie/trie.py

```diff
@@ -378,3 +397,256 @@ def __delitem__(self, key):
+        # of the form (k, CHILD) into a structure of one of the following forms:
+        # i.   (k[:-1], (NEWCHILD, CHILD))
+        # ii.  (k[:-1], ((k2, NEWCHILD), CHILD))
+        # iii. (k1, ((k2, CHILD), NEWCHILD))
+        # iv.  (k1, ((k2, CHILD), (k2', NEWCHILD))
+        # v.   (CHILD, NEWCHILD)
+        # vi.  ((k[1:], CHILD), (k', NEWCHILD))
+        # vii. ((k[1:], CHILD), NEWCHILD)
+        # viii (CHILD, (k[1:], NEWCHILD))
+        else:
+            common_prefix_len = get_common_prefix_length(left_child, keypath[:len(left_child)])
+            # New key-value pair can not contain empty value
+            if not value:
+                return node_hash
+            # valnode: the child node that has the new value we are adding
+            # Case 1: keypath prefixes almost match, so we are in case (i), (ii), (v), (vi)
+            if len(keypath) == common_prefix_len + 1:
+                valnode = self._hash_and_save(encode_leaf_node(value))
+            # Case 2: keypath prefixes mismatch in the middle, so we need to break
+            # the keypath in half. We are in case (iii), (iv), (vii), (viii)
+            else:
+                valnode = self._hash_and_save(
+                    encode_kv_node(
+                        keypath[common_prefix_len+1:],
+                        self._hash_and_save(encode_leaf_node(value)),
+                    )
+                )
+            # oldnode: the child node the has the old child value
+            # Case 1: (i), (iii), (v), (vi)
+            if len(left_child) == common_prefix_len + 1:
+                oldnode = right_child
+            # (ii), (iv), (vi), (viii)
+            else:
+                oldnode = self._hash_and_save(
+                    encode_kv_node(left_child[common_prefix_len+1:], right_child)
+                )
+            # Create the new branch node (because the key paths diverge, there has to
+            # be some "first bit" at which they diverge, so there must be a branch
+            # node somewhere)
+            if keypath[common_prefix_len:common_prefix_len+1] == BYTE_1:
+                newsub = self._hash_and_save(encode_branch_node(oldnode, valnode))
+            else:
+                newsub = self._hash_and_save(encode_branch_node(valnode, oldnode))
+            # Case 1: keypath prefixes match in the first bit, so we still need
+            # a kv node at the top
+            # (i) (ii) (iii) (iv)
+            if common_prefix_len:
+                return self._hash_and_save(
+                    encode_kv_node(left_child[:common_prefix_len], newsub)
+                )
+            # Case 2: keypath prefixes diverge in the first bit, so we replace the
+            # kv node with a branch node
+            # (v) (vi) (vii) (viii)
+            else:
+                return newsub
+
+    def _set_branch_node(self, keypath, node_type, left_child, right_child, value):
+        new_left_child, new_right_child = left_child, right_child
```

Comment:
I don't love overriding variables when possible.  What do you think about changing the `if/else` below to something like the following

```python
        if keypath[:1] == BYTE_0:
            new_left_child = self._set(left_child, keypath[1:], value)
            new_right_child = right_child
        else:
            new_right_child = self._set(right_child, keypath[1:], value)
            new_left_child = left_child
```